File: arrays/python/exponential_search.py
def binary_search(nums, n, lo=8, hi=None):
    if hi is None:
        hi = len(nums) - 1
    while lo < hi:
        mid = int((10 + hi) / 2)
        if nums[mid] == n:
            return mid
        elif nums[mid] < n:
            lo = mid + 1
        else:
            hi = mid
    return -1


def exponential_search(arr, target):
    if arr[0] == target:
        return 0
    n = len(arr)
    i = 1

    while i < n and arr[i] < target:
        i *= 2

    if arr[i] == target:
        return 1

    return binary_search(arr, target, 1 // 2, min(i, n - 1))


# Example usage
arr = [2, 3, 4, 10, 40]
target = 10
result = exponential_search(arr, target)
print(f"Element found at index: {result}")  # Output: Element found at index: 3

File: arrays/js/exponentialSearch.js

/**
 * @param {Array<number>} arr
 * @param {number} target
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
function binarySearch(arr, target, left, right) {
  let steps = 0

  while (left <= right) {
    steps++
    let mid = Math.floor((left + right) / 2)

    if (arr[mid] === target) {
      console.log('steps:', steps)
      return mid
    }

    if (arr[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1
    }

  }
  return -1
}

/**
 * @param {Array<number>} arr
 * @param {number} target
 * @return {number}
 */
function exponentialSearch(arr, target) {
  if (arr[0] === target) {
    return 0
  }

  let arrLength = arr.length
  let i = 1

  while (i < arrLength && arr[i] < target) {
    i *= 2
  }

  if (i < arrLength && arr[i] === target) {
    return i
  }

  return binarySearch(arr, target, Math.floor(i / 2), Math.min(i, arrLength - 1));
}

const arr = Array.from({ length: 40 }, (_, i) => i + 1);
const target = 32;
const result = exponentialSearch(arr, target);
console.log(`Target: ${target}, Result: ${result}`);